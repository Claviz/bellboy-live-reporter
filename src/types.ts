import { IBellboyEvent } from "bellboy";

export interface IReporterConfig {
    url?: string;
}

export interface ReportData {
    events: IBellboyEvent[];
    reportFinished?: boolean;
}