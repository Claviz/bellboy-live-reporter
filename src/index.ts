import axios from 'axios';
import { IBellboyEvent, Job, Reporter } from 'bellboy';

import { IReporterConfig, ReportData } from './types';

const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
let savedEvents: IBellboyEvent[] = [];
let condition: string;

async function poll(url: string, reportData: ReportData) {
    try {
        const response = await axios.post(url, reportData);
        savedEvents = [];
        condition = response.data?.toLowerCase();
        return;
    } catch (err) {
        const timeToWait = 3000;
        console.log(`Can't reach ${url}`);
        console.log(`Waiting ${timeToWait / 1000}s before reconnect...`);
        await delay(timeToWait);
        await poll(url, reportData);
    }
}
class LogReporter extends Reporter {

    #url: string;

    constructor(config?: IReporterConfig) {
        super();
        this.#url = config?.url || '';
    }

    report(job: Job) {
        this.#url = this.#url ? this.#url : `http://localhost:3041`;
        job.onAny(undefined, async (event) => {
            savedEvents.push(event);
            const reportFinished = event.eventName === 'endProcessing';
            if (reportFinished || !condition || (condition && JSON.stringify(event).toLowerCase().includes(condition))) {
                await poll(this.#url, { events: savedEvents, reportFinished });
            }
        });
    }
}

export = LogReporter;