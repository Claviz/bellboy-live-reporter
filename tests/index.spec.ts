import { Destination, DynamicProcessor, IBellboyEvent, Job } from 'bellboy';
import express from 'express';
import { Server } from 'http';

import LiveReporter from '../src';
import { ReportData } from '../src/types';

let data: ReportData[] = [];

function startServer(opts?: { simulateFailOnce?: boolean; port?: number; }): Server {
    function mockChangingData(reportData: ReportData) {
        const changed = {
            ...reportData,
            events: reportData.events.map(x => ({ ...x, jobId, timestamp, eventId: `event${timestamp}` })),
        };
        timestamp += 1;

        return changed;
    }
    const jobId = 'job1';
    let timestamp = 0;
    let firstTime = true;
    const app = express();
    app.use(express.json());
    app.post('/', function (req: any, res: any) {
        const body = req.body as ReportData;
        data.push(mockChangingData(body));
        res.send(firstTime && opts?.simulateFailOnce ? 404 : undefined);
        firstTime = false;
    });
    app.post('/with-condition', function (req: any, res: any) {
        const body = req.body as ReportData;
        data.push(mockChangingData(body));
        res.send(firstTime && opts?.simulateFailOnce ? 404 : 'something_3');
        firstTime = false;
    });
    const server = app.listen(opts?.port ? opts.port : 3041);

    return server;
}

beforeAll(async () => {
});

beforeEach(async () => {
    data = [];
});

afterAll(async () => {
});

afterEach(async () => {
});

class CustomDestination extends Destination {
    async loadBatch(rows: any) {
        // do nothing
    }
}

it('successfully sends event requests to the server', async () => {
    const server = startServer();
    const processor = new DynamicProcessor({
        generator: async function* () {
            for (let i = 0; i < 3; i++) {
                yield {
                    text: 'something'
                }
            }
        },
    });
    const destination = new CustomDestination();
    const job = new Job(processor, [destination], { reporters: [new LiveReporter()] });
    await job.run();
    expect(data).toMatchSnapshot();
    server.close();
});

it('successfully sends event requests to the custom server endpoint', async () => {
    const port = 3042;
    const server = startServer({ port: port });
    const processor = new DynamicProcessor({
        generator: async function* () {
            for (let i = 0; i < 3; i++) {
                yield {
                    text: 'something'
                }
            }
        },
    });
    const destination = new CustomDestination();
    const job = new Job(processor, [destination], { reporters: [new LiveReporter({ url: `http://localhost:${port}` })] });
    await job.run();
    expect(data).toMatchSnapshot();
    server.close();
});

it('retries to send an event to the server on fail', async () => {
    const server = startServer({ simulateFailOnce: true });
    const processor = new DynamicProcessor({
        generator: async function* () {
            for (let i = 0; i < 3; i++) {
                yield {
                    text: 'something'
                }
            }
        },
    });
    const destination = new CustomDestination();
    const job = new Job(processor, [destination], { reporters: [new LiveReporter()] });
    await job.run();
    expect(data).toMatchSnapshot();
    server.close();
});

it('if condition is set, waits for it before sending events', async () => {
    const server = startServer();
    const processor = new DynamicProcessor({
        generator: async function* () {
            for (let i = 0; i < 5; i++) {
                yield {
                    text: `something_${i}`
                }
            }
        },
    });
    const destination = new CustomDestination();
    const job = new Job(processor, [destination], { reporters: [new LiveReporter({ url: `http://localhost:3041/with-condition` })] });
    await job.run();
    expect(data).toMatchSnapshot();
    server.close();
});