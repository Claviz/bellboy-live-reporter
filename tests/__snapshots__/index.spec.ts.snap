// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if condition is set, waits for it before sending events 1`] = `
Array [
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          Object {
            "rowLimit": 0,
          },
          Array [
            Object {
              "batchSize": 0,
              "disableLoad": false,
            },
          ],
        ],
        "eventId": "event0",
        "eventName": "startProcessing",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 0,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event1",
        "eventName": "startProcessingStream",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 1,
      },
      Object {
        "eventArguments": Array [
          Object {
            "text": "something_0",
          },
        ],
        "eventId": "event1",
        "eventName": "startProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 1,
      },
      Object {
        "eventArguments": Array [],
        "eventId": "event1",
        "eventName": "endProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 1,
      },
      Object {
        "eventArguments": Array [
          Object {
            "text": "something_1",
          },
        ],
        "eventId": "event1",
        "eventName": "startProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 1,
      },
      Object {
        "eventArguments": Array [],
        "eventId": "event1",
        "eventName": "endProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 1,
      },
      Object {
        "eventArguments": Array [
          Object {
            "text": "something_2",
          },
        ],
        "eventId": "event1",
        "eventName": "startProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 1,
      },
      Object {
        "eventArguments": Array [],
        "eventId": "event1",
        "eventName": "endProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 1,
      },
      Object {
        "eventArguments": Array [
          Object {
            "text": "something_3",
          },
        ],
        "eventId": "event1",
        "eventName": "startProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 1,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event2",
        "eventName": "endProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 2,
      },
      Object {
        "eventArguments": Array [
          Object {
            "text": "something_4",
          },
        ],
        "eventId": "event2",
        "eventName": "startProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 2,
      },
      Object {
        "eventArguments": Array [],
        "eventId": "event2",
        "eventName": "endProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 2,
      },
      Object {
        "eventArguments": Array [
          0,
          Array [
            Object {
              "text": "something_0",
            },
            Object {
              "text": "something_1",
            },
            Object {
              "text": "something_2",
            },
            Object {
              "text": "something_3",
            },
            Object {
              "text": "something_4",
            },
          ],
        ],
        "eventId": "event2",
        "eventName": "loadingBatch",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 2,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          0,
          Array [
            Object {
              "text": "something_0",
            },
            Object {
              "text": "something_1",
            },
            Object {
              "text": "something_2",
            },
            Object {
              "text": "something_3",
            },
            Object {
              "text": "something_4",
            },
          ],
        ],
        "eventId": "event3",
        "eventName": "loadedBatch",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 3,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          0,
        ],
        "eventId": "event4",
        "eventName": "endLoadingBatch",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 4,
      },
      Object {
        "eventArguments": Array [],
        "eventId": "event4",
        "eventName": "endProcessingStream",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 4,
      },
      Object {
        "eventArguments": Array [],
        "eventId": "event4",
        "eventName": "endProcessing",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 4,
      },
    ],
    "reportFinished": true,
  },
]
`;

exports[`retries to send an event to the server on fail 1`] = `
Array [
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          Object {
            "rowLimit": 0,
          },
          Array [
            Object {
              "batchSize": 0,
              "disableLoad": false,
            },
          ],
        ],
        "eventId": "event0",
        "eventName": "startProcessing",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 0,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          Object {
            "rowLimit": 0,
          },
          Array [
            Object {
              "batchSize": 0,
              "disableLoad": false,
            },
          ],
        ],
        "eventId": "event1",
        "eventName": "startProcessing",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 1,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event2",
        "eventName": "startProcessingStream",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 2,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          Object {
            "text": "something",
          },
        ],
        "eventId": "event3",
        "eventName": "startProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 3,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event4",
        "eventName": "endProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 4,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          Object {
            "text": "something",
          },
        ],
        "eventId": "event5",
        "eventName": "startProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 5,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event6",
        "eventName": "endProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 6,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          Object {
            "text": "something",
          },
        ],
        "eventId": "event7",
        "eventName": "startProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 7,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event8",
        "eventName": "endProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 8,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          0,
          Array [
            Object {
              "text": "something",
            },
            Object {
              "text": "something",
            },
            Object {
              "text": "something",
            },
          ],
        ],
        "eventId": "event9",
        "eventName": "loadingBatch",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 9,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          0,
          Array [
            Object {
              "text": "something",
            },
            Object {
              "text": "something",
            },
            Object {
              "text": "something",
            },
          ],
        ],
        "eventId": "event10",
        "eventName": "loadedBatch",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 10,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          0,
        ],
        "eventId": "event11",
        "eventName": "endLoadingBatch",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 11,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event12",
        "eventName": "endProcessingStream",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 12,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event13",
        "eventName": "endProcessing",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 13,
      },
    ],
    "reportFinished": true,
  },
]
`;

exports[`successfully sends event requests to the custom server endpoint 1`] = `
Array [
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          Object {
            "rowLimit": 0,
          },
          Array [
            Object {
              "batchSize": 0,
              "disableLoad": false,
            },
          ],
        ],
        "eventId": "event0",
        "eventName": "startProcessing",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 0,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event1",
        "eventName": "startProcessingStream",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 1,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          Object {
            "text": "something",
          },
        ],
        "eventId": "event2",
        "eventName": "startProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 2,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event3",
        "eventName": "endProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 3,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          Object {
            "text": "something",
          },
        ],
        "eventId": "event4",
        "eventName": "startProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 4,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event5",
        "eventName": "endProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 5,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          Object {
            "text": "something",
          },
        ],
        "eventId": "event6",
        "eventName": "startProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 6,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event7",
        "eventName": "endProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 7,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          0,
          Array [
            Object {
              "text": "something",
            },
            Object {
              "text": "something",
            },
            Object {
              "text": "something",
            },
          ],
        ],
        "eventId": "event8",
        "eventName": "loadingBatch",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 8,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          0,
          Array [
            Object {
              "text": "something",
            },
            Object {
              "text": "something",
            },
            Object {
              "text": "something",
            },
          ],
        ],
        "eventId": "event9",
        "eventName": "loadedBatch",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 9,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          0,
        ],
        "eventId": "event10",
        "eventName": "endLoadingBatch",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 10,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event11",
        "eventName": "endProcessingStream",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 11,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event12",
        "eventName": "endProcessing",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 12,
      },
    ],
    "reportFinished": true,
  },
]
`;

exports[`successfully sends event requests to the server 1`] = `
Array [
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          Object {
            "rowLimit": 0,
          },
          Array [
            Object {
              "batchSize": 0,
              "disableLoad": false,
            },
          ],
        ],
        "eventId": "event0",
        "eventName": "startProcessing",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 0,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event1",
        "eventName": "startProcessingStream",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 1,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          Object {
            "text": "something",
          },
        ],
        "eventId": "event2",
        "eventName": "startProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 2,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event3",
        "eventName": "endProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 3,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          Object {
            "text": "something",
          },
        ],
        "eventId": "event4",
        "eventName": "startProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 4,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event5",
        "eventName": "endProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 5,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          Object {
            "text": "something",
          },
        ],
        "eventId": "event6",
        "eventName": "startProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 6,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event7",
        "eventName": "endProcessingRow",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 7,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          0,
          Array [
            Object {
              "text": "something",
            },
            Object {
              "text": "something",
            },
            Object {
              "text": "something",
            },
          ],
        ],
        "eventId": "event8",
        "eventName": "loadingBatch",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 8,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          0,
          Array [
            Object {
              "text": "something",
            },
            Object {
              "text": "something",
            },
            Object {
              "text": "something",
            },
          ],
        ],
        "eventId": "event9",
        "eventName": "loadedBatch",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 9,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [
          0,
        ],
        "eventId": "event10",
        "eventName": "endLoadingBatch",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 10,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event11",
        "eventName": "endProcessingStream",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 11,
      },
    ],
    "reportFinished": false,
  },
  Object {
    "events": Array [
      Object {
        "eventArguments": Array [],
        "eventId": "event12",
        "eventName": "endProcessing",
        "jobId": "job1",
        "jobStopped": false,
        "timestamp": 12,
      },
    ],
    "reportFinished": true,
  },
]
`;
